[ParseableInterface] Honor "exported module names"

jrose-apple opened the pull request:
Clang's "exported module name" is a pretty obscure feature (and one we wish we didn't need), but sometimes API is initially exposed through one module in order to build another one, and we want the canonical presented name to be something else. Push this concept into Swift's AST properly, then use that name when printing a parseable interface, just like we do with serialization.

rdar://problem/49114811

jrose-apple committed:
[ClangImporter] Avoid unneeded copy for getExportedModuleName

std::string -> StringRef of long-lived backing storage.

No functionality change.

jrose-apple committed:
Push Clang's "exported module name" notion up to Swift's FileUnit

It's a pretty obscure feature (and one we wish we didn't need), but
sometimes API is initially exposed through one module in order to
build another one, and we want the canonical presented name to be
something else. Push this concept into Swift's AST properly so that
other parts of the compiler stop having to know that this is a
Clang-specific special case.

No functionality change in this commit; will be used in the next
commit.

jrose-apple commented:
@swift-ci Please test

harlanhaskins commented on test/ParseableInterface/exported-module-name.swift:
```diff
@@ -0,0 +1,19 @@
+// RUN: %empty-directory(%t)
+// RUN: %target-swift-frontend -typecheck -emit-parseable-module-interface-path %t/CoreKitClient.swiftinterface -module-name CoreKitClient -I %S/Inputs/exported-module-name-before %s
```
I think `exported-module-name-before` wasn't committed?

jrose-apple replied:
Fixed, thanks.

jrose-apple committed:
[ParseableInterface] Honor "exported module names"

(as described in the previous commit)

When printing an interface that has to be stable, we need to use the
module name that identifies where declarations should be searched for,
just like we do with serialization.

rdar://problem/49114811

jrose-apple commented:
@swift-ci Please test

harlanhaskins reviewed (approved):
Thanks for fixing this, and good catch!

compnerd commented:
This caused a regression on the Windows tests!

jrose-apple commented:
:-( What was it? Did I do something too shell-y again?
